feat: DID Exchange Inbound Message Event

rolsonquadras opened the pull request:
- Support to register action events(go channels) for inbound DIDExchange message types
- Support to register for auto execution of message processing
- Support to register stream events(go channels) for inbound DIDExchange message types
- Support for callback for action events


Closes #198 

Signed-off-by: Rolson Quadras <rolson.quadras@securekey.com>

llorllale commented on pkg/didcomm/protocol/didexchange/service.go:
```diff
@@ -64,18 +96,138 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
+
+// processCallback processes the callback events.
+func (s *Service) processCallback(didCommCallback event.DIDCommCallback) {
+	// TODO what to do in case of any failures with callback processing ? - currently, logging the error messages and
```
"What do do?"

Probably communicate the error via `DIDCommEvent` (adding an "Err" attribute maybe?)?

llorllale replied:
Or perhaps we should have a separate stream of errors encountered.

rolsonquadras replied:
Created a issue for the TODO https://github.com/hyperledger/aries-framework-go/issues/242

llorllale commented on pkg/didcomm/event/event.go:
```diff
@@ -0,0 +1,21 @@
+import "github.com/hyperledger/aries-framework-go/pkg/didcomm/dispatcher"
+
+// DIDCommEvent message type to pass events in go channels.
+type DIDCommEvent struct {
```
I think both these new structs need more extensive godocs, including their attributes.

Eg. clearly specify what `DIDCommCallback.Err` does.

rolsonquadras replied:
[Updated](https://github.com/hyperledger/aries-framework-go/compare/35ffd11efb21de8bceb7b468e53a833a2b1a8f82..e20d32d8200881f1bfdc2f36aad1988fa5a78e93#diff-d3e84998ff09ae2ecd4128d1239a97e7R24).

llorllale commented on pkg/didcomm/protocol/didexchange/service.go:
```diff
@@ -64,18 +96,138 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
 }
 
+// Register events on DID Exchange protocol.
+func (s *Service) Register(msgType string, ch chan<- event.DIDCommEvent) (chan<- event.DIDCommCallback, error) {
```
Looks like we need to export the msgTypes as well.

rolsonquadras replied:
Valid point. [Updated](https://github.com/hyperledger/aries-framework-go/compare/35ffd11efb21de8bceb7b468e53a833a2b1a8f82..e20d32d8200881f1bfdc2f36aad1988fa5a78e93#diff-f761c91e753eba1c42a68076df06ce12R34).

rolsonquadras commented on pkg/didcomm/protocol/didexchange/service.go:
```diff
@@ -40,11 +57,31 @@ type provider interface {
 func New(store storage.Store, prov provider) *Service {
-	return &Service{outboundTransport: prov.OutboundTransport(), store: store}
+	// create channels for callback
+	// TODO configure callback channel size
```
Issue #246

bstasyszyn commented on pkg/didcomm/protocol/didexchange/service.go:
```diff
@@ -63,21 +85,60 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
+	}
+
+	if _, ok := s.events[msgType]; ok {
+		s.events[msgType] = append(s.events[msgType], ch)
```
I think that you'll need a mutex around s.events.

rolsonquadras replied:
[Added](https://github.com/hyperledger/aries-framework-go/compare/2fc8ca5f292de18ad5657f1f1317e24252476a3c..9e40655340de2e0f5c4fc9f2074c4f8cd3e1126d#diff-f761c91e753eba1c42a68076df06ce12R61) mutex.

bstasyszyn commented on pkg/didcomm/protocol/didexchange/service_test.go:
```diff
@@ -478,30 +486,354 @@ func (p *mockProvider) OutboundTransport() transport.OutboundTransport {
+				require.Equal(t, ConnectionInvite, e.Message.Type)
+
+				// invoke callback
+				err = e.Callback(event.DIDCommCallback{Err: nil})
```
When Callback is invoked, processCallback will be invoked and handle(...) will now run in this Go routine (i.e. the client's Go routine). This Go routine will be blocked until the flow has completed which may result in deadlock if the client needs to handle more events.

rolsonquadras replied:
@bstasyszyn Very valid point. I was thinking to reduce the work for consumers to call the go-channel as they had to map the correlation keys.

For this, may be we can start a go routine or invoke channel inside call back function. So that it can be returned the control back immediately. If this is not a blocker for now, can we tackle this issue in next PRs/issue ? I can create the issue for this.

rolsonquadras replied:
[Wrapped](https://github.com/hyperledger/aries-framework-go/compare/9e40655340de2e0f5c4fc9f2074c4f8cd3e1126d..686bbecbf5c870012cfbab30d759addb432f52e9#diff-f761c91e753eba1c42a68076df06ce12R207) internal go channel inside the callback function for non-blocking callback from the consumer.

llorllale commented on pkg/didcomm/protocol/didexchange/service.go:
```diff
@@ -63,21 +85,60 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
+		return fmt.Errorf("event registration not supported for the message type %s", msgType)
+	}
+
+	if _, ok := s.events[msgType]; ok {
```
I'm suggesting the statement on line 114 is sufficient for both cases - when the key `msgType` already exists or when it does not.

rolsonquadras replied:
Got it, Thanks. Updated in the new [Patch](https://github.com/hyperledger/aries-framework-go/compare/2fc8ca5f292de18ad5657f1f1317e24252476a3c..9e40655340de2e0f5c4fc9f2074c4f8cd3e1126d#diff-f761c91e753eba1c42a68076df06ce12R116).

rolsonquadras commented:
1. Remove go routine - (https://github.com/hyperledger/aries-framework-go/pull/240/files#diff-f761c91e753eba1c42a68076df06ce12R179)
2. Non-block processCallback
3. Have a  errorEvent
4. Client deals with one channel (event). 
5. Hide all other async details from clients. eg - callback appears to be blocking to the client even though its non-blocking.

troyronda commented on pkg/didcomm/protocol/didexchange/service.go:
```diff
@@ -63,21 +87,58 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
+}
+
+// Register events on DID Exchange protocol.
+func (s *Service) Register(msgType string, ch chan<- event.DIDCommEvent) error {
```
This function is currently conflating five activities:

1. Register a Go channel to listen to the event stream.
2. Register a Go channel to listen to workflow steps that require an action.
3. Set which events require an explicit callback action to progress the state machine.
4. Allow a different Go channel for each event type.
5. Filter the events that are returned to the client.

I argue that:
- (3) should have been done prior to the protocol starting.
- (4) seems unnecessary.
- (5) I could see some arguments for filtering... but I think the caller code looks the same even with filtering... and this is all local (not filtering a remote stream)...

troyronda replied:
Also what is hidden behind the pattern of passing a channel as an argument is that:
- the caller can control the channel buffer size.
- the caller can use the same channel against multiple protocols.

rolsonquadras commented:
> 1. Remove go routine - (https://github.com/hyperledger/aries-framework-go/pull/240/files#diff-f761c91e753eba1c42a68076df06ce12R179)
> 2. Non-block processCallback
> 3. Have a  errorEvent
> 4. Client deals with one channel (event).
> 5. Hide all other async details from clients. eg - callback appears to be blocking to the client even though its non-blocking.

Addressed all the above comments except point#3 (TODO #242 )

rolsonquadras commented on pkg/didcomm/protocol/didexchange/service.go:
```diff
@@ -86,6 +147,75 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
 
+func (s *Service) sendEvent(msg dispatcher.DIDCommMsg, threadID string, nextState state) (bool, error) {
+	invoked := false
+	if _, ok := s.events[msg.Type]; ok {
```
noted.

troyronda commented on pkg/didcomm/protocol/didexchange/service.go:
```diff
@@ -39,11 +61,22 @@ type provider interface {
 func New(store storage.Store, prov provider) *Service {
-	return &Service{outboundTransport: prov.OutboundTransport(), store: store}
+	svc := &Service{
+		outboundTransport: prov.OutboundTransport(),
```
There should be a TODO that we actually need an outbound dispatcher rather than the transport directly.

The dispatcher needs to call Pack prior to using the outbound transport.

rolsonquadras replied:
[Added](https://github.com/hyperledger/aries-framework-go/compare/6ca38f34c66c2017dee287d5ecd2734ca169e7b7..5839b0f509b0932a67abd562fb6a05dc181f2e08#diff-f761c91e753eba1c42a68076df06ce12R73) a TODO. Issue #259.

troyronda commented on pkg/didcomm/protocol/didexchange/service.go:
```diff
@@ -39,11 +61,22 @@ type provider interface {
 type Service struct {
 	outboundTransport transport.OutboundTransport
 	store             storage.Store
+	ch                chan didCommChMessage
```
Suggest a more descriptive name.

rolsonquadras replied:
[Updated](https://github.com/hyperledger/aries-framework-go/compare/6ca38f34c66c2017dee287d5ecd2734ca169e7b7..5839b0f509b0932a67abd562fb6a05dc181f2e08#diff-f761c91e753eba1c42a68076df06ce12R66).

troyronda commented on pkg/didcomm/protocol/didexchange/service.go:
```diff
@@ -86,6 +150,92 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
+func (s *Service) startInternalListener() {
+	go func() {
+		for msg := range s.ch {
+			// TODO handle error in callback - https://github.com/hyperledger/aries-framework-go/issues/242
```
Also need to ensure we have tests for:

- no listeners call the callback.
- multiple listeners call the callback.
- some listeners call the callback and some do not.

troyronda commented on pkg/didcomm/protocol/didexchange/service.go:
```diff
@@ -63,21 +96,52 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
+// Register events on DID Exchange protocol. The events are triggered for incoming ConnectionRequest, ConnectionResponse
+// or ConnectionAck message types.
+func (s *Service) Register(ch chan<- event.DIDCommEvent) error {
+	s.event = append(s.event, ch)
```
Is this thread safe?

rolsonquadras replied:
No. As per I remember during the last discussion, there were comments that registration happens when protocol is initiated. Hence locks are not necessary as such and having it will be an overhead during fetching the registered event to trigger it. Since we don't have that code now, I have added the mutex now.

Note - I had the locks earlier and removed it in previous patch.

troyronda replied:
I don't know how registrations can happen when the protocol is initiated.

A client package is created "on the fly".

troyronda commented on pkg/didcomm/protocol/didexchange/service.go:
```diff
@@ -63,21 +96,52 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
+
+// Register events on DID Exchange protocol. The events are triggered for incoming ConnectionRequest, ConnectionResponse
+// or ConnectionAck message types.
+func (s *Service) Register(ch chan<- event.DIDCommEvent) error {
```
How does a listener unregister?

rolsonquadras replied:
I was thinking about adding this in second cut as this PR is already more than 600 lines. For now added an [empty function](https://github.com/hyperledger/aries-framework-go/compare/5839b0f509b0932a67abd562fb6a05dc181f2e08..5dfa12128f1ba2867b1a8c7c47464e6f9b8c5b7a#diff-f761c91e753eba1c42a68076df06ce12R134) with a TODO.

troyronda commented on pkg/didcomm/protocol/didexchange/service.go:
```diff
@@ -12,24 +12,46 @@ import (
+	ConnectionAck = DIDExchangeSpec + "ack"
 )
 
+type message struct {
```
It would be good to have a comment on how message is being used.

rolsonquadras replied:
[Updated](https://github.com/hyperledger/aries-framework-go/compare/6ca38f34c66c2017dee287d5ecd2734ca169e7b7..5839b0f509b0932a67abd562fb6a05dc181f2e08#diff-f761c91e753eba1c42a68076df06ce12R50)

troyronda commented on pkg/didcomm/protocol/didexchange/service.go:
```diff
@@ -63,21 +96,52 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
+	// trigger the event if the message type is registered (if it's not outbound)
+	eventsTriggered := false
+	if !msg.Outbound {
+		eventsTriggered, err = s.sendEvent(msg, thid, next)
```
I assume eventsTriggered is only set to true for events that have callback actions (e.g., not error events).

rolsonquadras replied:
At the moment, events are triggered only for connectionRequest, response and ack. [Ref](https://github.com/hyperledger/aries-framework-go/pull/240/files#diff-f761c91e753eba1c42a68076df06ce12R162)

troyronda commented on pkg/didcomm/protocol/didexchange/service.go:
```diff
@@ -63,21 +96,52 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
+	return nil
+}
+
+// Register events on DID Exchange protocol. The events are triggered for incoming ConnectionRequest, ConnectionResponse
```
There is an undocumented side-effect: auto execution when you don't register; auto execution if you do.

Also as a TODO: you might only be registering for events not for callbacks. Related to the thought above: https://github.com/hyperledger/aries-framework-go/pull/240/files#r324795675

rolsonquadras replied:
Updated the Godoc and created a TODO. #261

troyronda commented on pkg/didcomm/protocol/didexchange/service.go:
```diff
@@ -63,21 +96,52 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
+
+	// if no events are trigger continue the execution
+	if !eventsTriggered {
+		return s.handle(&message{Msg: msg, ThreadID: thid, NextStateName: next.Name()})
```
We have to be careful here.

There is a race between starting the protocol and disabling auto execution (currently via register). We don't want to accidentally auto execute state and we don't want to accidentally drop messages either :).

rolsonquadras replied:
valid point. Created #261 .

rolsonquadras replied:
This has been addressed in this PR. Godoc - https://github.com/hyperledger/aries-framework-go/pull/240/files#diff-f761c91e753eba1c42a68076df06ce12R133.

Baha-sk commented on pkg/didcomm/protocol/didexchange/service.go:
```diff
@@ -192,8 +342,12 @@ func (s *Service) marshalAndSend(data interface{}, errorMsg, destination string)
 }
 
+func (s *Service) generateRandomID() string {
+	return uuid.New().String()
```
not sure why, but the uuid docs says "creates new UUID or panics": https://godoc.org/github.com/google/uuid#New

it's maybe rare, but we should probably not allow panics here.. or at least catch them.

generally speaking, for random bytes, I prefer using crypto.Rand
have you tried  uuid.NewRandom() ?

troyronda replied:
Note uuid.New and uuid.NewRandom are [equivalent](https://github.com/google/uuid/blob/master/version4.go#L14) except for the panic vs error return.


So the following blurb from NewRandom also applies to New.
> The strength of the UUIDs is based on the strength of the crypto/rand

rolsonquadras replied:
@Baha-sk just curious about the error logic for such cases. Whats the error/retry logic processing in such cases (error, panic)? Should we throw error back to the consumers or retry till we succeed?

troyronda replied:
The panic happens if the application [cannot read](https://github.com/google/uuid/blob/master/version4.go#L37) from crypto/rand. This seems a bit unrecoverable...

Baha-sk commented on pkg/didcomm/protocol/didexchange/service_test.go:
```diff
@@ -478,30 +487,352 @@ func (p *mockProvider) OutboundTransport() transport.OutboundTransport {
+}
+
+func randomString() string {
+	u := uuid.New()
```
same here

rolsonquadras replied:
Discussion is tracked [here](https://github.com/hyperledger/aries-framework-go/pull/240#discussion_r324806191).

rolsonquadras commented:
Updated the code based on discussion with @troyronda. 

Configs:
1. Action Events - Consumer can register one event for message processing. The event message will have a callback function. The consumer need to call this to continue processing the message.
2. Stream Events -  Consume can register multiple event for stream listening. This is not related to message handling and just a mean for the consumers to listen to the messages.
3. Auto Execution - Consumers can set Auto execution to execute the message processing. 

Behaviour:
1. If Action events or auto execution not set - the service will throw error while handling the messages.
2. If action event set after auto execution - the service will trigger action event and wait for callback
3. If action event set before auto execution - the service will auto execute
3. If multiple action event set - the events will triggered on last event channel.

troyronda commented on cmd/aries-agentd/main_test.go:
```diff
@@ -103,7 +103,7 @@ func validateRequests(t *testing.T) {
 							}`),
 				"application/didcomm-envelope-enc"),
-			expectedStatus:     http.StatusAccepted,
+			expectedStatus:     http.StatusInternalServerError,
```
Is this the status for the protocol doesn't yet have action callback listeners?

There seems to be a couple of relevant error codes to choose from: https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#5xx_Server_errors

e.g., 503

rolsonquadras replied:
@troyronda yes. The error codes are set in [inbound.go](https://github.com/hyperledger/aries-framework-go/blob/master/pkg/didcomm/transport/http/inbound.go#L62).

troyronda replied:
Ah. We should consider a TODO where we have a more specific code like 503.

rolsonquadras replied:
#271

troyronda commented on pkg/didcomm/protocol/didexchange/service.go:
```diff
@@ -63,21 +112,115 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
+func (s *Service) RegisterActionEvent(ch chan<- event.DIDCommEvent) error {
+	s.lock.Lock()
+	s.actionEvent = ch
+	s.execute = true
```
is execute meaning service is allowed to respond?

rolsonquadras replied:
yes.

llorllale commented on cmd/aries-agentd/main_test.go:
```diff
@@ -103,7 +103,7 @@ func validateRequests(t *testing.T) {
 							}`),
 				"application/didcomm-envelope-enc"),
-			expectedStatus:     http.StatusAccepted,
+			expectedStatus:     http.StatusInternalServerError,
```
why this change?

rolsonquadras replied:
[Here](https://github.com/hyperledger/aries-framework-go/pull/240#discussion_r325327735).

troyronda commented on pkg/didcomm/protocol/didexchange/service.go:
```diff
@@ -63,21 +112,115 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
+// RegisterStreamEvent on DID Exchange protocol messages. The events are triggered for incoming ConnectionRequest,
+// ConnectionResponse or ConnectionAck message types. The Callback is set to nil in the actionEvent message and service
+// won't be expecting any response from the consumers.
+func (s *Service) RegisterStreamEvent(ch chan<- dispatcher.DIDCommMsg) error {
```
RegisterStatusEvent (or similar) might be the better corollary to RegisterActionEvent

troyronda replied:
You may also want to match up the type names.

If you are calling the type DIDCommMsg then I would expect this method to be RegisterMsg[Foo].
Similarly with RegisterActionEvent returning DIDCommEvent then I would expect that method to be RegisterEvent[Foo].

rolsonquadras replied:
Updated.
1. RegisterActionEvent to [RegisterEvent](https://github.com/hyperledger/aries-framework-go/compare/6dc9c7de378434ea27520364556f649ccf434c5f..36d8a92bfaa82b830e44a047521b376920c82c72#diff-f761c91e753eba1c42a68076df06ce12R143)
2. RegisterStreamEvent to [RegisterMsg](https://github.com/hyperledger/aries-framework-go/compare/6dc9c7de378434ea27520364556f649ccf434c5f..36d8a92bfaa82b830e44a047521b376920c82c72#diff-f761c91e753eba1c42a68076df06ce12R160)

troyronda commented on pkg/didcomm/protocol/didexchange/service.go:
```diff
@@ -63,21 +112,115 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
+}
+
+// UnregisterActionEvent on DID Exchange protocol messages. Refer RegisterActionEvent().
+func (s *Service) UnregisterActionEvent() error {
```
I'd expect the channel to be passed into such a method.

rolsonquadras replied:
Since it's one channel(event), what are we expecting to do with that arg ?

troyronda commented on pkg/didcomm/protocol/didexchange/service.go:
```diff
@@ -63,21 +112,115 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
+
+// RegisterActionEvent on DID Exchange protocol messages. The events are triggered for incoming ConnectionRequest,
+// ConnectionResponse or ConnectionAck message types. The consumer need to invoke the callback to resume processing.
+// Only one channel can be registered for the action events. If called multiple times, the events will be sent to the
```
The alternative, of course, is to return an error if currently registered.

troyronda commented on pkg/didcomm/protocol/didexchange/service.go:
```diff
@@ -192,8 +430,12 @@ func (s *Service) marshalAndSend(data interface{}, errorMsg, destination string)
 	return s.outboundTransport.Send(string(jsonString), destination)
 }
 
+func (s *Service) generateRandomID() string {
```
I don't see a reference to s in this method.

rolsonquadras replied:
[Updated](https://github.com/hyperledger/aries-framework-go/compare/6dc9c7de378434ea27520364556f649ccf434c5f..36d8a92bfaa82b830e44a047521b376920c82c72#diff-f761c91e753eba1c42a68076df06ce12R437).

llorllale commented on pkg/didcomm/protocol/didexchange/service.go:
```diff
@@ -63,21 +112,115 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
+// RegisterStreamEvent on DID Exchange protocol messages. The events are triggered for incoming ConnectionRequest,
+// ConnectionResponse or ConnectionAck message types. The Callback is set to nil in the actionEvent message and service
+// won't be expecting any response from the consumers.
+func (s *Service) RegisterStreamEvent(ch chan<- dispatcher.DIDCommMsg) error {
```
`dispatcher.DIDCommMsg` is not an event...

llorllale replied:
Also naming is everything. This method's purpose would be a lot clearer with a name such as "Register**For**EventsStream" or something similar. Keyword: for.

rolsonquadras replied:
Discussion is tracker [here](https://github.com/hyperledger/aries-framework-go/pull/240#discussion_r325329926).

llorllale commented on pkg/didcomm/protocol/didexchange/service.go:
```diff
@@ -63,21 +112,115 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
+}
+
+// UnregisterAutoExecute on DID Exchange protocol messages. Refer RegisterAutoExecute().
+func (s *Service) UnregisterAutoExecute() error {
```
This should not disable the events stream.

troyronda replied:
I assume you mean DIDCommMsg as a result of RegisterStreamEvent :). Agree those should not be disabled.

rolsonquadras replied:
@llorllale @troyronda Is the code missing something here? Does this need any change ?

bstasyszyn commented on pkg/didcomm/protocol/didexchange/service.go:
```diff
@@ -39,15 +64,39 @@ type provider interface {
 // Handle didexchange msg
 func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
+	// throw error if there are no action events are registered or auto execute set (if it's not outbound)
+	if !msg.Outbound && !s.execute {
```
Not sure if this is related to the comment by Troy below, but s.execute needs an RLock in order to read.

rolsonquadras replied:
[Added](https://github.com/hyperledger/aries-framework-go/compare/6dc9c7de378434ea27520364556f649ccf434c5f..36d8a92bfaa82b830e44a047521b376920c82c72#diff-f761c91e753eba1c42a68076df06ce12R97).

llorllale commented on pkg/didcomm/protocol/didexchange/service.go:
```diff
@@ -39,15 +64,39 @@ type provider interface {
 // Handle didexchange msg
 func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
+	// throw error if there are no action events are registered or auto execute set (if it's not outbound)
+	if !msg.Outbound && !s.execute {
```
This is not checking whether "action events" have been registered. If there are, then the msg should be processed, even if auto-execute is disabled.

rolsonquadras replied:
s.execute is derived based on actionEvent or AutoExecute registration.

llorllale commented on pkg/didcomm/protocol/didexchange/service.go:
```diff
@@ -39,15 +64,39 @@ type provider interface {
 type Service struct {
 	outboundTransport transport.OutboundTransport
 	store             storage.Store
+	callbackChannel   chan didCommChMessage
```
Suggest use of "inbound/outbound" or "incoming/outgoing" prefixes to add clarity:

* callbackChannel -> inboundCallbacks
* actionEvent -> outboundActions
* streamEvents -> outboundEvents

`execute` could use a rename to `autoExecute`

troyronda replied:
execute doesn't mean autoExecute (which also means it could use a rename to something more explicit.)

llorllale commented on pkg/didcomm/protocol/didexchange/service.go:
```diff
@@ -39,15 +64,39 @@ type provider interface {
+		execute: false,
+	}
+
+	svc.startInternalListener()
```
Suggest rename to "startListeningForIncomingCallbacks". At least it makes the intent more clear.

rolsonquadras replied:
@troyronda thoughts on this?

troyronda replied:
I think we need to do further changes as a new PR. It's getting hard to follow.

llorllale commented on pkg/didcomm/protocol/didexchange/service.go:
```diff
@@ -63,21 +112,115 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
+// last channel. This works in conjunction with RegisterAutoExecute(). If this function is called after
+// RegisterAutoExecute(), the service will not execute the processing automatically and it'll wait for the callback. If
+// both are not set, then service will not handle the messages and throw an error.
+func (s *Service) RegisterActionEvent(ch chan<- event.DIDCommEvent) error {
```
Same here w.r.t. naming - suggest "RegisterForEventsStream" (the other one should stay with "RegisterForMsgStream")

troyronda replied:
I would drop "For"

rolsonquadras replied:
Discussion is tracker [here](https://github.com/hyperledger/aries-framework-go/pull/240#discussion_r325329926).

llorllale replied:
@troyronda the intent is to register a channel via which events would be received by the caller of this function.

`RegisterEventStream(chan event.DIDCommEvent)` reads as though whoever is calling this function is going to produce the events.

`RegisterForEventStream(chan event.DIDCommEvent)` removes that ambiguity.

Also, "Stream" is probably redundant. Instead, just say "Events" - the "stream" is implicit.

Just my 2c.
